gitreceived/gitreceived.go

```diff
@@ -307,14 +313,134 @@ func ensureCacheRepo(path string) error {
 	defer cacheMtx.Unlock()
 
 	cachePath := *repoPath + "/" + path
-	if _, err := os.Stat(cachePath); os.IsNotExist(err) {
-		os.MkdirAll(cachePath, 0755)
+
+	// remove old copy
+	if _, err := os.Stat(cachePath); err == nil {
+		os.RemoveAll(cachePath)
+	}
+	os.MkdirAll(cachePath, 0755)
+
+	res, err := http.Get("http://blobstore.discoverd/cache/" + path + ".tar")
+	if err != nil {
+		return err
+	}
+	defer res.Body.Close()
+
+	if res.StatusCode == 200 {
+		// cache hit
+		r := tar.NewReader(res.Body)
+		for {
+			header, err := r.Next()
+			if err != nil {
+				if err == io.EOF {
+					break
+				}
+				return err
+			}
+
+			filename := cachePath + "/" + header.Name
+
+			switch header.Typeflag {
+			case tar.TypeDir:
+				if _, err := os.Stat(filename); os.IsNotExist(err) {
+					if err := os.MkdirAll(filename, os.FileMode(header.Mode)); err != nil {
+						return err
+					}
+				}
+				if err := os.Chmod(filename, os.FileMode(header.Mode)); err != nil {
+					return err
+				}
+			case tar.TypeReg, tar.TypeRegA:
+				// if the files are out of order, the dir might not exist yet
+				if _, err := os.Stat(filepath.Dir(filename)); os.IsNotExist(err) {
+					if err := os.MkdirAll(filepath.Dir(filename), os.FileMode(header.Mode)); err != nil {
+						return err
+					}
+				}
+				writer, err := os.Create(filename)
+				if err != nil {
+					return err
+				}
+				defer writer.Close()
+
+				io.Copy(writer, r)
+
+				if err := os.Chmod(filename, os.FileMode(header.Mode)); err != nil {
```

Comment:
Use OpenFile to set the mode and create at the same time: http://golang.org/pkg/os/#OpenFile